test/distributed/_tensor/test_attention.py

```diff
@@ -0,0 +1,250 @@
+        (
+            out_chunk,
+            logsumexp_chunk,
+            *others,
+        ) = _scaled_dot_product_flash_attention_chunked(
+            query,
+            key,
+            value,
+            size=self.world_size,
+            is_causal=is_causal,
+        )
+
+        out_chunk.sum().backward()
+        self.assertEqual(
+            out_chunk.shape, (bs, nheads, self.world_size * query_tokens, dim)
+        )
+        self.assertEqual(logsumexp_chunk, logsumexp)
+        self.assertEqual(out_chunk, out)
+        out_chunk_grad = query.grad
+        query.grad = None
+        # gradient doesn't match due to numerical issues with chunk size > 1
+        # self.assertEqual(out_chunk_grad, out_grad)
+
+        # parallel behavior
+        with attention_context_parallel(), CommDebugMode() as comm_mode:
+            (
+                out_parallel,
+                logsumexp_parallel,
+                *others,
+            ) = torch.ops.aten._scaled_dot_product_flash_attention(
+                dquery, dkey, dvalue, is_causal=is_causal
+            )
+        self.assertDictEqual(
+            comm_mode.get_comm_counts(),
+            {
+                c10d_functional.all_to_all_single: self.world_size - 1,
+            },
+        )
+        self.assertEqual(out_parallel.placements, (Shard(2),))
+        self.assertEqual(
+            out_parallel._local_tensor.shape, (bs, nheads, query_tokens, dim)
+        )
+        self.assertEqual(
+            out_parallel.shape, (bs, nheads, self.world_size * query_tokens, dim)
+        )
+        out_parallel_tensor = out_parallel.full_tensor()
+        self.assertEqual(out_parallel_tensor, out)
+        logsumexp_parallel_tensor = logsumexp_parallel.full_tensor()
+        self.assertEqual(logsumexp_parallel_tensor, logsumexp)
+
+        self.assertIsNone(dquery.grad)
+        with attention_context_parallel(), CommDebugMode() as comm_mode:
+            out_parallel.sum().backward()
+
+        self.assertDictEqual(
+            comm_mode.get_comm_counts(),
+            {
+                c10d_functional.all_to_all_single: (self.world_size - 1) * 2,
+            },
+        )
+        out_parallel_grad = dquery.grad.full_tensor()
+        dquery.grad = None
+        self.assertEqual(out_parallel_grad, out_chunk_grad)
+
+    @skip_if_lt_x_gpu(2)
+    @unittest.skipIf(
+        not PLATFORM_SUPPORTS_FLASH_ATTENTION, "Does not support flash attention"
+    )
+    @with_comms
+    @attention_context_parallel()
+    @sdpa_kernel(backends=[SDPBackend.FLASH_ATTENTION])
+    @parametrize("is_causal", [True, False])
+    def test_ring_attention_transformer(self, is_causal: bool):
+        device_mesh = DeviceMesh(
+            self.device_type,
+            torch.arange(0, self.world_size),
+        )
+        dtype = torch.bfloat16
+        bs = 8
+        ntokens = 8
+        dim = 32
+        nheads = 8
+        num_layers = 2
+
+        encoder_layer = nn.TransformerEncoderLayer(
```

Comment:
I wonder instead of testing it with `nn.TransformerEncoderLayer`, shall we test it directly with a custom TransformerLayer instead (i.e. we have some custom transformer code in https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/distributed/_tensor/common_dtensor.py#L144)

The reason is more like the Transformer nn.Module is less used in OSS and most of users built their own Transformer, so not sure how reliable the former one could be

Reply:
I think it's useful to test both here -- we want to have the native layers work out of the box. I added an additional test for the custom Transformer as it forced this implementation to generalized better to self-attention only style operators.